test/hotspot/jtreg/containers/cgroup/NestedCgroup.java

```diff
@@ -0,0 +1,268 @@
+                throw new SkippedException("cgroup/cgroup2 filesystem mount point not found");
+            }
+            sysFsCgroup = matcher.group(1);
+            System.err.println("sysFsCgroup = " + sysFsCgroup);
+
+            System.err.println(LINE_DELIM + " " + (isCgroup2 ? "cgroup2" : "cgroup1") + " mount point: " + sysFsCgroup);
+            memory_max_filename = isCgroup2 ? "memory.max" : "memory.limit_in_bytes";
+            Files.writeString(Path.of(sysFsCgroup + "/" + CGROUP_OUTER + "/" + memory_max_filename), "" + MEMORY_MAX_OUTER);
+
+            // Here starts a copy of ProcessTools.createJavaProcessBuilder.
+            List<String> cgexec = new ArrayList<>();
+            Limits limits = hook(cgexec);
+            OutputAnalyzer output = pSystem(cgexec);
+            // C++ CgroupController
+            output.shouldMatch("\\[trace\\]\\[os,container\\] Memory Limit is: " + limits.integer + "$");
+
+            pSystem(cgdelete);
+        }
+
+        public abstract Limits hook(List<String> cgexec) throws IOException;
+    }
+    private static class TestTwoLimits extends Test {
+        public Limits hook(List<String> cgexec) throws IOException {
+            // CgroupV1Subsystem::read_memory_limit_in_bytes considered hierarchical_memory_limit only when inner memory.limit_in_bytes is unlimited.
+            Files.writeString(Path.of(sysFsCgroup + "/" + CGROUP_OUTER + "/" + CGROUP_INNER + "/" + memory_max_filename), "" + MEMORY_MAX_INNER);
+
+            args_add_cgexec(cgexec);
+            args_add_self_verbose(cgexec);
+            cgexec.add("-version");
+
+            // KFAIL - verify the CgroupSubsystem::initialize_hierarchy() and jdk.internal.platform.CgroupSubsystem.initializeHierarchy() bug
+            // TestTwoLimits does not see the lower MEMORY_MAX_OUTER limit.
+            Limits limits = new Limits();
+            limits.integer = MEMORY_MAX_INNER;
+            return limits;
+        }
+        public TestTwoLimits() throws Exception {
+        }
+    }
+    private static class TestNoController extends Test {
+        public Limits hook(List<String> cgexec) throws IOException {
+            args_add_cgexec(cgexec);
+            args_add_self(cgexec);
+            cgexec.add("NestedCgroup");
+            cgexec.add("TestNoController");
+            cgexec.add(Test.jdkTool);
+            cgexec.add(sysFsCgroup + "/" + CGROUP_OUTER + "/cgroup.subtree_control");
+
+            Limits limits = new Limits();
+            limits.integer = MEMORY_MAX_OUTER;
+            return limits;
+        }
+        public TestNoController() throws Exception {
+        }
+        public static void child(String arg) throws Exception {
+            Files.writeString(Path.of(arg), "-memory");
+
+            List<String> self_verbose = new ArrayList<>();
+            args_add_self_verbose(self_verbose);
+            self_verbose.add("-version");
+            pSystem(self_verbose);
+        }
```

Comment:
Instead of disabling the memory controller that way, consider creating only the cpu controller (using `cgcreate`) and test for memory limits?